### ISSUE TITLE
Alarm delayed when the screen is off for a while

### ISSUE THREAD

**Issue by Sesu8642** [Sesu8642/InfusionTimer#12]
I noticed this when brewing a tea bag for a couple of minutes. The bell sound only happened once I turned on the screen. This is probably caused by some energy saving feature similar to #8.

**Comment by Sesu8642** [comment]
It can be fixed by setting "alarmClock: true" for the alarm. However, this causes Android to display a visual indication that there is an active alarm. I would prefer not having the indication but there doesn't seem to be another way.

**Comment by theAkito** [comment]
Android is getting stricter and stricter about background services, etc.

At this time, there is no way around having a visual indicator, to make sure the background job keeps running.

And even then, it may sometimes still be killed, if the OS sees fit.

On Xiaomi and Samsung devices with stock ROMs it's even worse than that. They sometimes even kill system services to "save energy" and artificially keep the battery run for a bit longer.

**Comment by Sesu8642** [comment]
What an annoyance... Thanks for confirming!

**Comment by geronime** [comment]
Hello there, after the latest Android updade I also don't get the alarm when the phone locked itself in the meantime. This is quite unpleasant with longer brews as then I completely forget about the infusion in progress...
(Google Pixel 5, stock Android 12, March 5 security update)

**Comment by Sesu8642** [comment]
@geronime this is very interesting. I cant explain how the change I did could have worsened the problem. Does the alarm happen once you turn on the screen? And do you see the visual indication that an alarm is scheduled?

**Comment by geronime** [comment]
> Does the alarm happen once you turn on the screen?
Yep.

Funny fact: Now the phone is connected to the power adapter and the alarm sounded even in locked state!

**Comment by Sesu8642** [comment]
I was able to reproduce the problem again both on versions 1.2.0 and 1.2.1 using a 5 minute timer.
I did a little analysis on the problem. I use the Android Alarm Manager Plus plugin for scheduling the alarm:
```
AndroidAlarmManager.oneShotAt(infusionFinishTime, ALARM_ID, _ring,
allowWhileIdle: true,
exact: true,
wakeup: true,
alarmClock: true); // since version 1.2.1
```
The parameters are evaluated in the scheduleAlarm function in the [AlarmService](https://github.com/fluttercommunity/plus_plugins/blob/main/packages/android_alarm_manager_plus/android/src/main/java/dev/fluttercommunity/plus/androidalarmmanager/AlarmService.java) of the plugin.

Without alarmClock (like in version 1.2.0), the resulting call should be [AlarmManagerCompat.setExactAndAllowWhileIdle](https://developer.android.com/reference/android/app/AlarmManager#setExactAndAllowWhileIdle(int,%20long,%20android.app.PendingIntent)) with the [RTC_WAKEUP](https://developer.android.com/reference/android/app/AlarmManager#RTC_WAKEUP) type.
However, if alarmClock is set to true (like in version 1.2.1), all the other options are ignored and [AlarmManagerCompat.setAlarmClock](https://developer.android.com/reference/androidx/core/app/AlarmManagerCompat#setAlarmClock(android.app.AlarmManager,long,android.app.PendingIntent,android.app.PendingIntent)) is called. This is apparently like exact + wakeup + visual indication. Possibly allowWhileIdle as well, see [here](https://developer.android.com/reference/android/app/AlarmManager#setAlarmClock(android.app.AlarmManager.AlarmClockInfo,%20android.app.PendingIntent)), but I am unsure about this one.

Anyway, neither of the alternatives work properly. There needs to be something I am missing.

**Comment by Sesu8642** [comment]
One more observation: I cannot reproduce the issue while debugging 🥳
This is probably because the device is charging while connected to the PC... And its probably the reason why I thought that I fixed it earlier.

**Comment by Sesu8642** [comment]
@geronime I found that disabling battery optimization works as a workaround (at least for me). I am still looking for a proper fix though.

**Comment by Sesu8642** [comment]
I did some more testing and disabling battery optimization does not fully solve the problem. But it does reduce the delay significantly.

Then I tested with [Simple Clock](https://github.com/SimpleMobileTools/Simple-Clock) and it had the same problems I had. At the time, the app was abandoned. The stated reason was that the app was not reliable enough... Luckily, it looks like it is now maintained again.

After that, I tested with [Deskclock](https://github.com/qw123wh/Aospdeskclock) which was accurate and reliable on all of my devices. The app uses a CPU wake lock and displays a notification while the timer is running.

My goal with InfusionTimer is to be as accurate as Deskclock. I now use [Flutter Background](https://github.com/JulianAssmann/flutter_background) which:
- displays a notification
- acquires a CPU wake lock
- asks the user for the permission to run in the background

On my devices, the timer is now reliable. However, they are all running AOSP roms. One drawback of this is the additional notification which I don't think is necessary because there already is the progress notification but there is no option to turn it off in the Flutter Backgtound plugin.

**TLDR:**
To schedule an accurate alarm on Android, you only need to use the Alarm Manager with 3 separate parameters that all tell it to be accurate. In addition to that, you need to keep your app running all the time by acquiring a CPU wake lock, displaying a notification and annoying the user by getting their permission. Yes, it's as easy as that! :D
And will it work on stock roms? Who knows...

**Comment by Sesu8642** [comment]
Commit: eb71b002405ef439923481af286ce1a2bb6ad72d

**Comment by Sesu8642** [comment]
@geronime Would you mind testing a pre-release on your device? And if not, do you have a way to backup and restore the app's data (like Titanium or [Neo Backup](https://f-droid.org/en/packages/com.machiav3lli.backup/))?

**Comment by geronime** [comment]
Unfortunately I don't have Neo Backup (I don't even have f-droid installed on my Pixel 5, just stock Android, and from what I read on the linked page it requires rooted device).

Maybe we could ask @JaromirDvoracek to test on his Graphene OS?

I don't mind testing a pre-release as long as I can export/import my tea configurations - I would prefer not to loose my settings anymore. Maybe an idea for another feature - some plain export/import functionality, also for sharing between friends?

**Comment by Sesu8642** [comment]
Yes, these backup solutions do need a rooted device.

An integrated export/import feature does make sense for cases like this. I also thought about sharing before but that is probably way more complicated.

**Comment by geronime** [comment]
Okay, actually I can test it now already - @JaromirDvoracek recommended to create an extra user which has a separate space for apps so I don't need to worry about messing up my saved teas.
Which version do you want me to test?
Btw. the import/export feature would be still appreciated, eventually 🙂

**Comment by Sesu8642** [comment]
Fantastic :D
I am uploading the pre-release here because I don't want F-Droid to push it to all users again. It is based on the latest commit on master ([b81d1c9](https://github.com/Sesu8642/InfusionTimer/commit/b81d1c99f0406a973b809fa519e30292f5f2895a)).
[InfusionTimer 1.3.0-rc.1.zip](https://github.com/Sesu8642/InfusionTimer/files/8695721/InfusionTimer.1.3.0-rc.1.zip)

**Comment by geronime** [comment]
![Screenshot_20220516-084239](https://user-images.githubusercontent.com/1003724/168534642-07f3c63f-38dd-41cc-b4c6-80358f0ddf1f.png)
Hmm... The different user profile on the phone is not as separated as I thought... Could you please name the app differently in the pre-release (e.g. "Beta" at the end)?

**Comment by Sesu8642** [comment]
@geronime this one should be installable alongside the other one.
[InfusionTimer_changed_package 1.3.0-rc.1.zip](https://github.com/Sesu8642/InfusionTimer/files/8732841/InfusionTimer_changed_package.1.3.0-rc.1.zip)

**Comment by geronime** [comment]
Okay, that works! The search feature from #14 looks great!
I was prompted to allow unrestricted battery usage. Quick test shows that the sound notification happens even with the display turned off but then again my battery is way above 50 % at the moment and the alarm was more likely to be delayed when the battery was in the bottom half. I tested the default western style brewing white tea and both infusions, second one being 4 minutes, gave me the sound notification even though I explicitly turned the display off right after starting the timer.
I'll do some more tests once my battery is low and let you know.

**Comment by geronime** [comment]
Tested with battery below 25 % crosschecking with a countdown on another device and the alarm goes off on time. ✅

**Comment by Sesu8642** [comment]
Thank you for testing this thoroughly! :)
I will publish the release soon.
